imsApp/src/imsRecord.cc

```diff
@@ -3589,6 +3647,142 @@ static void post_fields( imsRecord *prec, unsigned short alarm_mask,
     return;
 }
 
+/******************************************************************************/
+static void io_controller ( struct ims_info *mInfo )
+{
+    char *s, *t, buf[2048];
+    const double wtimeout = 1.0;
+    const double rtimeout = 0.1;
+    asynStatus status;
+    int i, eomReason, wild;
+    size_t len;
+
+    for (;;) {
+	/* Check for new work */
+	mInfo->rMutex->lock();
+	for (i = 0; i < REQ_CNT; i++) {
+	    struct ims_asyn_req *req = &mInfo->req[i];
+
+	    if (req->state == IAR_SENDONLY) {
+		pasynOctetSyncIO->write( mInfo->pasynUser, req->out, 
+					 req->outsize, wtimeout, &len );
+		free_request_buf(mInfo, req);
+		continue;
+	    }
+	    if (req->state == IAR_SEND) {
+		pasynOctetSyncIO->write( mInfo->pasynUser, req->out, 
+					 req->outsize, wtimeout, &len );
+		req->state = IAR_RESPWAIT;
+		continue;
+	    }
+	}
+	mInfo->rMutex->unlock();
+
+	/* Try to read a message */
+	status = pasynOctetSyncIO->read( mInfo->pasynUser, buf, sizeof(buf) - 1,
+					 rtimeout, &len, &eomReason );
+	if (status != asynSuccess || len <= 0)
+	    continue;
+
+	if (len < sizeof(buf))
+	    buf[len] = 0;
+
+#if 0
+	/* Debug print. */
+	printf("< ");
+	for (size_t ii = 0; ii < len; ii++) {
+	    if (buf[ii] == '\r')
+		printf("\\r");
+	    else if (buf[ii] == '\n')
+		printf("\\n");
+	    else
+		putchar(buf[ii]);
+	}
+	putchar('\n');
+#endif
+
+	/* Is it even possible to have more than one message?!? */
+	for (s = buf, t = index(s, '\r'); s && *s; s = t) {
+	    t = index(s, '\r');
+	    if (t) {
+		*t++ = 0;
+		if (*t == '\n')
+		    t++;
+	    }
+	    /* 
+	     * OK, now s is a single message string.  Check if we
+	     * need to handle it here.
+	     */
+	    if (!strncmp(s, "BOS", 3)) {
+		/*
+		 * BOSstatus,P=positionEOS
+		 * Strip off "BOS" and pass it to $(MOTOR).SSTR.
+		 */
+		if (!mInfo->sstr_status)
+		    dbPutField(&mInfo->sstr_addr, DBR_STRING, s+3, 1);
+		continue;
+	    }
+	    if (!strncmp(s, "Want2Save", 9)) {
+		/*
+		 * Want2Save
+		 * Strip off "Want2" and pass it to $(MOTOR).SVNG.
+		 */
+		if (!mInfo->svng_status)
+		    dbPutField(&mInfo->svng_addr, DBR_STRING, s+5, 1);
+		continue;
+	    }
+	    if (!strncmp(s, "Saved ", 6)) {
+		/*
+		 * Saved position
+		 * Strip off "Saved " and pass it to $(MOTOR).SVNG and
+		 * $(MOTOR).SAVED_STR.
+		 */
+		if (!mInfo->svng_status)
+		    dbPutField(&mInfo->svng_addr, DBR_STRING, s+6, 1);
+		if (!mInfo->saved_status)
+		    dbPutField(&mInfo->saved_addr, DBR_STRING, s+6, 1);
+		continue;
+	    }
+	    mInfo->rMutex->lock();
+	    /*
+	     * Check if this is a response to something previously sent.
+	     */
+	    for (wild = -1, i = 0; i < REQ_CNT; i++) {
+		struct ims_asyn_req *req = &mInfo->req[i];
```

Comment:
MAJOR NIT: the `auto` keyword is one of the coolest new features in since c11 (iirc). It makes things a lot more readable, the heuristic I made up for myself: "auto should be used when its obvious what type something should be and cleans up code readability; do not use auto just cause you're lazy". 

Apologies for so many nits :p

Reply:
I've never really been a fan of auto.  But I suppose that's just because I'm old. :-)